test/utils.test.ts

```diff
@@ -248,4 +249,224 @@ describe('utils', () => {
+
+    it('arrayTypeAndSize int[7]', () => {
+      const [elemTypeName, arraySize] = arrayTypeAndSize("int[7]");
+      expect(elemTypeName).to.equal('int')
+      expect(arraySize).to.includes.members([7])
+    })
+
+    it('arrayTypeAndSize int[2][3]', () => {
+
+      const [elemTypeName, arraySize] = arrayTypeAndSize("int[2][3]");
+      expect(elemTypeName).to.equal('int')
+      expect(arraySize).to.includes.members([2, 3])
+    })
+
+    it('arrayTypeAndSize bool[2][3][8][1]', () => {
+
+      const [elemTypeName, arraySize] = arrayTypeAndSize("bool[2][3][8][1]");
+      expect(elemTypeName).to.equal('bool')
+      expect(arraySize).to.includes.members([2, 3, 8, 1])
+    })
+
+  })
+
+
+  describe('checkArray()', () => {
+
+    it('checkArray int[3]', () => {
+      expect(checkArray([3, 3, 3], ['int', [3]])).to.true;
+    })
+
+    it('checkArray int[3]', () => {
+      expect(checkArray([3, 3], ['int', [3]])).to.false;
+    })
+
+    it('checkArray int[3]', () => {
+      expect(checkArray([3, 3, 1, 3], ['int', [3]])).to.false;
+    })
+
+    it('checkArray int[3]', () => {
+      expect(checkArray([3, 3, new Int(2)], ['int', [3]])).to.true;
+    })
+
+    it('checkArray int[3]', () => {
+      expect(checkArray([3, 3, new Bool(true)], ['int', [3]])).to.false;
+    })
+
+
+    it('checkArray int[2][3]', () => {
+      expect(checkArray([[3, 3, 3], [3, 12, 3]], ['int', [2, 3]])).to.true;
+    })
+
+    it('checkArray int[2][3]', () => {
+      expect(checkArray([[3, 3, 3], [3, 12, 3], [1,1,1]], ['int', [2, 3]])).to.false;
+    })
+
+    it('checkArray int[2][3]', () => {
+      expect(checkArray([[3, 3, 3], [3, 12]], ['int', [2, 3]])).to.false;
+    })
+
+    it('checkArray int[2][1][3]', () => {
+      expect(checkArray([[[3, 3, 3]], [[3, 12, 3]]], ['int', [2, 1, 3]])).to.true;
+    })
+
+    it('checkArray int[2][1][3]', () => {
+      expect(checkArray([[[3, 3, 3], 1], [[3, 12, 3]]], ['int', [2, 1, 3]])).to.false;
+    })
+
+
+    it('checkArray int[2][2][3]', () => {
+      expect(checkArray([[[3, 3, 3], [3, 3, 3]], [[3, 12, 3], [3, 3, 3]]], ['int', [2, 2, 3]])).to.true;
+    })
+
+
+    it('checkArray int[2][2][3]', () => {
+      expect(checkArray([[[3, 3, 3], [3, 3, 3]], [[3, 12, 3], [3, 3, 3]]], ['int', [2, 2, 3]])).to.true;
+    })
+
+
+    it('checkArray int[2][3][4]', () => {
+      expect(checkArray([ [
+        [1, 2, 3, 4],
+        [5, 6, 7, 8],
+        [9, 10, 11, 12]
+        ],
+        [
+        [13, 14, 15, 16],
+        [17, 18, 19, 20],
+        [21, 22, 23, 24]
+        ]], ['int', [2, 3, 4]])).to.true;
+    })
+
+    it('checkArray int[2][3][4]', () => {
+      expect(checkArray([[1, 3, 5], [2, 4, 6]], ['int', [2, 3, 4]])).to.false;
+    })
+
+  })
+
+
+  describe('flatternArray()', () => {
+    it('flatternArray int[2][2][3]', () => {
+      expect(flatternArray([[[3, 3, 3], [3, 3, 3]], [[3, 12, 3], [3, 3, 3]]])).to.includes.members([3, 3, 3, 3, 3, 3, 3, 12, 3, 3, 3, 3])
```

Comment:
Use distinct array elements. Otherwise u may miss some bug. The extreme example would be all 3s.
`expect(flatternArray([[[3, 3, 3], [3, 3, 3]], [[3, 3, 3], [3, 3, 3]]])).to.includes.members([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3])`

Use sth like
`expect(flatternArray([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]])).to.includes.members([])`